content/blog/copilot-lessons/index.md

```diff
@@ -0,0 +1,131 @@
+
+1. Determines the user wants to analyze an update
+2. Identifies which API to call
+3. Calls the API
+4. Summarizes the results
+
+But we already know the user's intent - they clicked a debug button. Having the LLM  figure out what the user wants wastes time and money when we already know. So, we directly call the analysis API to get the results and use the LLM solely for what it does best: summarizing technical output into clear, actionable explanations.
+
+This is another small win for our "Software Engineering over Prompt Engineering" approach. Traditional code handles the predictable parts, while AI focuses on the human-facing explanations.
+
+But while minimizing the LLM workload helped with efficiency, we soon faced an even trickier challenge: the deceptive polish of AI-generated outputs.
+
+## The Illusion of Correctness
+
+![Before and After](false-info.png)
+
+Large language models excel at generating well-structured, grammatically correct output. They makes neat tables, tell good stories, and generally sound confident. That's what makes them dangerous because this polished presentation can mask underlying flaws in the information itself, creating a false sense of confidence for users.
+
+One of our early testers, Pablo, encountered this firsthand. He posed a query to Pulumi Copilot, asking for a summary of resources within a specific project. The response he received was impeccably formatted, neatly categorizing resources by type and providing counts for each. It *looked* right, and for us humans sometimes looking right carries a lot of weight.
+
+However, a closer inspection revealed the numbers were way off. Copilot had asked for the wrong data and then summarized it beautifully - but incorrectly. This highlighted our next challenge: how do you systematically test a system that can be confidently wrong while sounding completely right?
+
+## Testing the Untestable: Validating LLM Outputs
+
+When testing traditional code, we expect deterministic outputs. With LLMs, even successful outcomes can vary significantly. Here's how we tackle this challenge.
+
+Our initial approach focused on keyword checking. For example, when testing our update analysis feature, we initially had a specific test case where a security error occurred. Our test validates that the LLM's response mentions "security" and describes the specific error condition.
+
+This worked for basic validation but quickly showed its limitations. For instance, when a user asked, "How many Lambdas am I running?" the LLM correctly identified the AWS Lambda resources but didn't use the word "running." failing the eval.
+
+These early failures highlighted the brittleness of keyword checking and the need for a more nuanced approach. Inspired by platforms like LangSmith, we started leveraging LLMs themselves for validation. We adopted Promptfoo for automated testing. Our test suite now runs against every code change, validating both response content and format.
```

Comment:
I am not sure we should list LangSmith and Promptfoo here.  We should just say we use the LLM to help us validate our evals.

Reply:
I'm rewording a bit. Let me know if that will work.